### PR TITLE
[Merged by Bors] - TY-2407 ffi support for URI/URL

### PR DESCRIPTION
**Pull request by rustonaut**
Add ffi support for dart Uri <-> rust Url types.

Required by #105/[TY-2342](https://xainag.atlassian.net/browse/TY-2342).

### CODE REVIEW

File: discovery_engine/lib/src/ffi/types/string.dart [hunk 1]
```diff
@@ -13,25 +13,46 @@
...
-    return utf8.decode(data);
+  ) =>
+      BoxedStr.fromRawParts(
+        ffi.get_string_buffer(place),
+        ffi.get_string_len(place),
+      ).readNative();
+}
+
+class BoxedStr {
+  final Pointer<Uint8> start;
```
Comment (acrrd) [inline discovery_engine/lib/src/ffi/types/string.dart:38]: `ptr` is a more common name and is less confusing.

File: discovery_engine/lib/src/ffi/types/string.dart [hunk 2]
```diff
@@ -13,25 +13,46 @@
...
+  final int len;
+
+  BoxedStr.fromRawParts(this.start, this.len);
+
+  /// Creates a `Box<str>` based on given dart string.
+  factory BoxedStr.create(String string) {
+    final utf8Bytes = utf8.encode(string);
+    final len = utf8Bytes.length;
+    final start = ffi.alloc_uninitialized_bytes(len);
```
Comment (acrrd) [inline discovery_engine/lib/src/ffi/types/string.dart:47]:
```suggestion
    final ptr= ffi.alloc_uninitialized_bytes(len);
```

File: discovery_engine_core/bindings/src/types/string.rs [hunk 3]
```diff
@@ -14,40 +14,31 @@
...
-/// - `len <= capacity` must hold
-/// - all bytes up to the new len must be initialized
-/// - the string buffer from index `0` to `len` must contain
-///   a valid utf8 string after the len was set.
+/// - It must be valid to write a `String` instance to given pointer,
+///   the pointer is expected to point to uninitialized memory.
+/// - The bytes `str_start..str_start+str_len` must be a sound rust string.
 #[no_mangle]
-pub unsafe extern "C" fn set_string_len(string: *mut String, len: usize) {
+pub unsafe extern "C" fn init_string_at(place: *mut String, str_start: *mut u8, str_len: usize) {
```
[the remark this answers is not in the document]
Comment (rustonaut) [reply discovery_engine_core/bindings/src/types/string.rs:39]: I will change it to `str_ptr`,`str_len` it's not the length or the uri, as parsing does include normalization.

File: discovery_engine_core/bindings/src/types/url.rs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+///
+/// # Safety
+///
+/// - The pointer must point to a sound initialized `Url` instance.
+#[no_mangle]
+pub unsafe extern "C" fn get_url_buffer(url: *const Url) -> *const u8 {
+    unsafe { &*url }.as_str().as_ptr()
+}
+
+/// Returns teh length of the `str` buffer in an `Url` instance.
```
[the remark this answers is not in the document]
Comment (rustonaut) [reply discovery_engine_core/bindings/src/types/url.rs:105]: I was wondering why it wasn't flagged, it turns out teh is a word with mostly the same meaning as the :joy:

**Review by acrrd — APPROVED** [review]

**Comment by rustonaut** [comment]
bors merge

**Pull request closed without merging**